lsif typed syntax highlighter v0.1

tjdevries opened the pull request:
@olafurpg use this PR instead. The other PR had some mess-ups and some other stuff from trying to update deps (but not happening in that PR anymore because @varungandhi-src is going to do some updates in another PR).

You should be able to just run this locally and then make requests against `localhost:8000/lsif` and get the `data` key of the JSON result with the bytes of the lsif proto. There was some problems with getting into JSON at this point due to the version of the protobuf crate we're using (could use an alpha one and get the JSON support, but we can chat about that later).

tjdevries commented:
btw, we need to regenerate the lsif and drop this in here before it gets merged, since we have in progress stuff happening with #29929

tjdevries commented:
btw, this is how we should do languages IMO: https://github.com/helix-editor/helix/tree/master/helix-syntax

varungandhi-src commented on docker-images/syntax-highlighter/queries/go/highlights.scm:
```diff
@@ -0,0 +1,200 @@
+(parameter_declaration (identifier) @parameter)
+(variadic_parameter_declaration (identifier) @parameter)
+
+((identifier) @constant
+ (#eq? @constant "_"))
+
+; ((identifier) @constant
+;  (#vim-match? @constant "^[A-Z][A-Z\\d_]+$"))
```
Leftovers from debugging?

tjdevries commented:
@olafurpg got rust bindings being generated. Add them to their own crate (could be published to crates.io if we wanted later so you could use these in repos outside of sg without exposing anything else).

tjdevries commented:
Also, squashed our commits because there was some goofiness in the commit history (or something, maybe from when you force pushed earlier and maybe I pulled weird). but all your changes should be here and i've rebased on main now as well

varungandhi-src commented on client/branded/src/global-styles/highlight.scss:
```diff
@@ -774,4 +774,114 @@
+        color: var(--hl-blue);
+        filter: brightness(105%);
+    }
+    .hl-typed-IdentifierNull {
+        color: var(--hl-orange);
+        filter: brightness(130%);
+    }
+    .hl-typed-BooleanLiteral {
+        color: var(--hl-orange);
+    }
+    .hl-typed-IdentifierConstant {
+        color: var(--hl-dark-blue-1);
+    }
+    .hl-typed-IdentifierMutableGlobal {
+        color: var(--hl-dark-blue-1);
+    }
```
Are these colors chosen for compatibility with the existing colors/is refining them a future action item? It seems a bit weird to have global mutable variables be highlighted the same way as constants. Similarly for highlighting `null` the same way as a boolean literal.

tjdevries replied:
moving to: https://github.com/sourcegraph/sourcegraph/issues/30995

efritz commented on .gitmodules:
```diff
@@ -0,0 +1,8 @@
+[submodule "tree-sitter-go"]
```
Is this ok? 😬

tjdevries replied:
what do you mean? Do you want a more specific name or something for the submodule? I don't think I can add gitmodules anywhere except at the root, so I'm not sure how else to add these as actual submodules besides this. It doesn't affect anything if no one does `git submodule update --init --recursive` style things

efritz replied:
Nah, is just _repo's first submodule_, I think. Would make sure there's no tooling concerns with @sourcegraph/dev-experience.

tjdevries replied:
Just to be clear, it doesn't need to be run / present / available / or anything for anyone just _running_ sourcegraph. You only need to have these pulled if you're planning on doing development on syntax-highlighter itself.

efritz replied:
CI will need it for builds, no?

tjdevries replied:
Ah, that's an interesting point. I'm not sure -- I will follow up on that today.

jhchabran replied:
Hey, thanks for the ping! Yes it does https://sourcegraph.com/github.com/sourcegraph/sourcegraph/-/blob/enterprise/dev/ci/internal/ci/operations.go?L619 so we're good here.

efritz replied:
@tjdevries I'm appeased if CI works without any changes outside this PR (or if they do if dev-ex signs off), and if there is a required git submodule init command required for using it locally that it's documented in the setup guide (and/or added to `sg check/doctor` cc @mrnugget).

efritz commented on schema/settings.schema.json:
```diff
@@ -197,6 +197,14 @@
           }
         },
+        "treeSitterEnabled": {
+          "description": "The maximum number of files a repo can have to use case-insensitive fuzzy finding",
```
🤔

efritz commented on client/web/src/lsif/spec.ts:
```diff
@@ -0,0 +1,109 @@
+    constructor(public readonly start: Position, public readonly end: Position) {}
+}
+
+// This is copy & pasted from the enum defined in lsif.proto.
+// Make sure to update it whenever you update lsif.proto
```
Can we enforce this in CI? (Or automate the copy+paste with a go-generate command? that automatically gets diffed in CI)

tjdevries replied:
The plan is once we figure out the dependency situation with protobuf (probably different PR) we will just directly import the typescript bindings and can delete this part.

efritz commented on client/branded/src/global-styles/highlight.scss:
```diff
@@ -774,4 +774,114 @@
+    }
+    .hl-typed-IdentifierBuiltin {
+        color: var(--hl-blue);
+        filter: brightness(105%);
```
Is this ... meant to be here?

tjdevries replied:
I removed for now, but I think it looks good :) it will be done in https://github.com/sourcegraph/sourcegraph/issues/30995 where we can talk just about CSS

efritz commented on docker-images/syntax-highlighter/docs/testing.md:
````diff
@@ -0,0 +1,15 @@
+Run tests with `$ cargo test`. If you have failures, you can do this cool thing:
+
+```
+$ cargo insta review
````
`insta`?

tjdevries replied:
yeah, it's a library you can install to run and manage snapshot tests.

efritz replied:
I honestly thought you forgot to type the rest of install.

efritz commented on docker-images/syntax-highlighter/docs/api.md:
```diff
@@ -0,0 +1,21 @@
+
+## `/lsif`
+
+TODO
```
- [ ] TODO

efritz commented:
LGMT atm. Let me know when you want a more formal approval (depending on which of the unchecked boxes in the PR description are pre-merge).

tjdevries commented on cmd/frontend/internal/highlight/highlight.go:
```diff
@@ -111,9 +121,9 @@ var ErrBinary = errors.New("cannot render binary file")
-func Code(ctx context.Context, p Params) (h template.HTML, aborted bool, err error) {
+func Code(ctx context.Context, p Params) (h template.HTML, l *lsif_typed.Document, aborted bool, err error) {
 	if Mocks.Code != nil {
-		return Mocks.Code(p)
```
@olafurpg do you know why you changed this (at least, I think it was you, I don't remember changing this part)

olafurpg replied:
This change should be reverted. I removed this line to get the code compiling after changing the signature. I tried running `./dev/generate.sh` but `Mocks.Code(p)` was still using the old signature.

tjdevries replied:
reverted

efritz commented on sg.config.yaml:
```diff
@@ -30,7 +30,8 @@ env:
   SYMBOLS_URL: http://localhost:3184
-  SRC_SYNTECT_SERVER: http://localhost:9238
+  # TODO: Don't merge, this is just for testing ease-of-use
+  SRC_SYNTECT_SERVER: http://localhost:8000
```
- [ ] Block

efritz commented on schema/settings.schema.json:
```diff
@@ -197,6 +197,14 @@
             "pointer": true
           }
         },
+        "treeSitterEnabled": {
```
Should this setting be scoped more specifically to highlighting?

efritz commented:
Approved assuming you check off the remainder of the blockers we've discussed/are marked in this PR. Tag me again if you'd like a focused re-review.

tjdevries commented on cmd/frontend/graphqlbackend/schema.graphql:
```diff
@@ -4346,6 +4351,10 @@ type HighlightedFile {
     """
+    Base64 encoded JSON payload of LSIF Typed with syntax highlighting data.
+    """
+    lsif: String!
```
@efritz  Do you think these are all the places that we should include this? I am not 100% familiar with our gql schema, so I'm happy to add / remove some of these if you think they shouldn't be here (or make them optional or something so that it's better for backwards compat or whatever)

tjdevries commented:
@efritz or @mrnugget do you know what this line is about: https://sourcegraph.com/gitlab.com/sourcegraph/sourcegraph/-/blob/cmd/server/Dockerfile?L61&subtree=true I'm just wondering if I need to update any hashes here (I'm not sure when this Dockerfile gets used / called, etc)

tjdevries commented:
@mrnugget Is it possible that we could just remove a hash from this entirely and it just uses whatever is the latest image?

https://sourcegraph.com/gitlab.com/sourcegraph/sourcegraph/-/blob/sg.config.yaml?L349

mrnugget commented:
> @efritz or @mrnugget do you know what this line is about: https://sourcegraph.com/gitlab.com/sourcegraph/sourcegraph/-/blob/cmd/server/Dockerfile?L61&subtree=true I'm just wondering if I need to update any hashes here (I'm not sure when this Dockerfile gets used / called, etc)

For context: that is the `Dockerfile` for the `server` image, which is a single Docker image that contains all of Sourcegraph. It's the image from our [one-line quickstart](https://docs.sourcegraph.com/#quick-install) and if I'm not mistaken it's also used by a bunch of customers (or users of the open-source version).

Going through the git history it looks like that image tag is updated by hand every time there was a new release of syntect. Example: https://github.com/sourcegraph/sourcegraph/commit/c9b9106fe2a8e3a042fbc8cb47a601e5cc5471f0

So if we want to switch to a new version of syntax-highlighter, then yes, we need to update that hash too. But maybe ask in #delivery or #dev-experience if someone knows when or how these hashes get updated.

> @mrnugget Is it possible that we could just remove a hash from this entirely and it just uses whatever is the latest image?
> https://sourcegraph.com/gitlab.com/sourcegraph/sourcegraph/-/blob/sg.config.yaml?L349

I just copy&pasted that line from the old dev scripts when creating the initial `sg.config.yaml`, I think.

For local dev I think it's fine to use `:latest` image, but then you should probably add a `docker pull` to the `install` (not `run`!) section for that entry.

kevinwojo commented:
Hey there @tjdevries, your friendly #delivery person here!

So the "magic" that drives the release of the syntax highlighter is in the [`build.sh`](https://github.com/sourcegraph/sourcegraph/commit/c9b9106fe2a8e3a042fbc8cb47a601e5cc5471f0#diff-bcbf8575d1369637986087f06dbd9931d40d87b72a4836e9a6b52f4d601c82f5) file.

From my understanding all of the necessary spots are the files identified the commit that @mrnugget found: https://github.com/sourcegraph/sourcegraph/commit/c9b9106fe2a8e3a042fbc8cb47a601e5cc5471f0

Confirming what @mrnugget says, it looks like this change was done manually each time. I'm not 100% sure on the "why" here. Perhaps @slimsag has some historical context here?

tjdevries commented:
Hmm, so is there a way for me to merge this commit and use the updated image at the same time? (I'm not sure if the question makes sense. Basically I have changes in both sourcegraph main repo and syntax-highlighter image and I want them to all be live at the same time.

Now, my understanding is for cloud it will use the latest image. So maybe I just need to do a follow up commit that updates the syntax-highlighter image to a new hash so that it works for our other docker images that we publish.

(Although I could be misunderstanding big time. I'm not familiar really with _how_ things get deployed. Just that they do :laughing: )

tjdevries commented:
Oh, also, after looking at that build.sh, I think it's quite out of date. The syntax-highlighter wasn't in-tree at the time of that commit, so it looks quite a bit different now. Now build.sh builds from the source tree in the monorepo.

@kevinwojo

tjdevries commented:
- [ ] Note: add hash back in for sg.config.yaml

olafurpg reviewed (approved):
LGTM 👍 I got this working locally and I'm super excited to see this go live! Really nice work @tjdevries 👏

mrnugget commented on client/web/src/repo/blob/backend.ts:
```diff
@@ -34,9 +41,10 @@ export const fetchBlob = memoizeObservable(
                     content
                     richHTML
-                    highlight(disableTimeout: $disableTimeout) {
+                    highlight(disableTimeout: $disableTimeout, treeSitterEnabled: $treeSitterEnabled) {
```
Any particular reason why you also didn't add this to `fetchHighlightedFileLineRanges`?

https://github.com/sourcegraph/sourcegraph/blob/58dbd909fd6fdba6a6be1229b2c536c888bb3404/client/web/src/repo/backend.ts#L192-L195

That's the function used to highlight search results.

mrnugget replied:
Sure, can be a follow-up. Just wanted to raise this since it means tree-sitter is only used when highlighting complete files, not search results, or (current) reference panel, etc.

mrnugget commented:
I added the `treeSitterEnabled: true` experimental feature flag to my settings (that should go in the test plan, btw.) but I can't get it to work:

<img width="1017" alt="screenshot_2022-02-22_11 19 38@2x" src="https://user-images.githubusercontent.com/1185253/155112282-6ec31f34-0cfd-430d-b927-c5cacafef0e2.png">

Log:

```
[enterprise-frontend] ERROR syntax highlighting failed (this is a bug, please report it), filepath: internal/database/external_services.go, filetype: Go, repo_name: github.com/sourcegraph/sourcegraph, revision: 2c26fb1acef3c230a85ee37a650d2223002e38aa, snippet: "package database\n\nimport (\n\t\"bytes\"\n\t\"context\"\n\t\"database/sql\"\n\t\"fmt\"\n\t\"net/url\""…, error: http://localhost:9238: gosyntect internal error: resource_not_found
```

I'm using `sg start enterprise-codeintel` so it should use the Docker image that you added, right?

mrnugget commented on .tool-versions:
```diff
@@ -11,3 +11,4 @@ trivy 0.20.0
 kustomize 4.0.5
 awscli 2.4.7
 python system
+rust 1.58.0
```
That's not the version used in the Dockerfile:

https://github.com/sourcegraph/sourcegraph/blob/58dbd909fd6fdba6a6be1229b2c536c888bb3404/docker-images/syntax-highlighter/Dockerfile#L4

I think it's better if they're consistent so I'd change one or the other.

mrnugget commented on docker-images/syntax-highlighter/README.md:
```diff
@@ -68,19 +50,19 @@ Once published, the image version will need to be updated in the following locat
-
 ## Adding themes
 
+TODO: Maybe we can just remove themes entirely. I think they are u
```
Create a ticket for that instead? Or at least finish the sentence 😄

tjdevries replied:
Yes, that already is this ticket: https://github.com/sourcegraph/sourcegraph/issues/31202 I want to ensure we're not actually using themes anywhere and then I'll delete it.

mrnugget commented on docker-images/syntax-highlighter/crates/sg-macros/src/lib.rs:
```diff
@@ -0,0 +1,64 @@
+};
+
+struct IncludeOptional {
+    // literals: Punctuated<LitStr, Token![,]>,
```
```suggestion
```

mrnugget reviewed (approved):
Pretty superficial review because it is a lot of code and a lot of pieces, but left a commend on rust-version that I think we should fix before merging.

Also: did you already push this to a `main-dry-run` branch to see if all the tests still pass? (see `sg ci build --help` for info)

camdencheek commented:
Changes here seem to be causing `dev/check/go-lint.sh` to fail with syntax errors on my machine:

```
--- golangci-lint
./dev/check/go-lint.sh: line 43: ./docker-images/syntax-highlighter: syntax error: operand expected (error token is "./docker-images/syntax-highlighter")
./dev/check/go-lint.sh: line 50: [: -v: unary operator expected
--- golangci-lint ./docker-images/syntax-highlighter
ERRO Running error: context loading failed: failed to load packages: failed to load with go/packages: err: exit status 1: stderr: go: no module declaration in go.mod. To specify the module path:
	go mod edit -module=example.com/mod 

./dev/check/go-lint.sh: line 33: ./annotations/go-lint: No such file or directory
```
